src/H5system.c

```
@@ -1456,3 +1456,124 @@ HDqsort_context(void *base, size_t nel, size_t size, int (*compar)(const void *,
+
+typedef struct HDqsort_fallback_context_t {
+    int (*gnu_compar)(const void *, const void *, void *);
+    void *gnu_arg;
+} HDqsort_fallback_context_t;
+
+#ifdef H5_HAVE_THREADSAFE
+/* Thread-local storage approach for thread-safe builds
+ *
+ * Note: The functions below use assertions instead of the HDF5 error stack because
+ * qsort and its variants return void, preventing propagation of errors.
+ */
+static H5TS_key_t  HDqsort_fallback_key;
+static H5TS_once_t HDqsort_fallback_key_once = H5TS_ONCE_INITIALIZER;
+
+static void
+HDqsort_fallback_key_init(void)
+{
+    herr_t ret = SUCCEED;
+
+    /* Create the thread-local storage key (no destructor needed) */
+    ret = H5TS_key_create(&HDqsort_fallback_key, NULL);
+
+    /* Assert that initialization succeeded - cannot propagate errors from here */
+    if (H5_UNLIKELY(ret < 0)) {
+        assert(false && "Failed to create TLS key for qsort fallback");
+        (void)0; /* Ensure non-empty body even when asserts are disabled */
+    }
+}
+
+static int
+HDqsort_fallback_wrapper(const void *a, const void *b)
+{
+    HDqsort_fallback_context_t *ctx = NULL;
+
+    /* Retrieve the context from thread-local storage
+     * This should never fail since we just set it in HDqsort_fallback() */
+    if (H5_UNLIKELY(H5TS_key_get_value(HDqsort_fallback_key, (void **)&ctx) < 0))
+        return 0; /* Should never happen, but return 0 (equal) if it does */
+
+    /* Call the original GNU-style comparator with context */
+    return ctx->gnu_compar(a, b, ctx->gnu_arg);
+}
+
+void
+HDqsort_fallback(void *base, size_t nel, size_t size, int (*compar)(const void *, const void *, void *),
+                 void *arg)
+{
+    HDqsort_fallback_context_t ctx;
+    herr_t                     ret;
+
+    /* Ensure the TLS key is initialized */
+    ret = H5TS_once(&HDqsort_fallback_key_once, HDqsort_fallback_key_init);
+    if (H5_UNLIKELY(ret < 0)) {
+        assert(false && "Failed to initialize TLS key for qsort fallback");
```

As this is in main library code, we need to do better here than asserting on a condition that has the potential to occur in debug builds and simply ignoring errors in release builds. Since the signature of this function currently doesn't allow propagating of errors, either the interface needs reworked to support that, or we should just abandon this variant of the function, using the global variable approach only, and document the thread-safety issues until we actually need a thread-safe version.